Spacing and bang operators

lorenzwalthert opened the pull request:
This closes #38 by adding two examples to section 2.2:

* spacing with `!!` and `!!!`.
* Note on operator precedence (https://github.com/tidyverse/style/issues/38#issuecomment-324155511). Example is taken from the `rlang` help file for `quasiquotation`.

The second point hasn't anything to do with spaces, so please advise if you want to list it under section 2.2 or elsewhere (or not at all), I just could not find a better place within the style guide.

hadley commented on 02-syntax.Rmd:
```diff
@@ -54,6 +54,34 @@ average <- mean(feet / 12 + inches, na.rm = TRUE)
 
+Put a space after the bang operators (`!!` and `!!!`) but not
+before them. Note that the sequence of exclamation marks must not be interrupted
+with spaces, otherwise the meaning is changed to the logical operator `!`.
```
It's not quite that the meaning is changed (at least from R's perspective). It's just confusing.

But it might be better to pull this out into a new chapter? (it'll be very short to start with). I'm worried that it's too esoteric to introduce here.

lorenzwalthert replied:
Or do you prefer to drop it completely?

lorenzwalthert replied:
I would not mind.

hadley replied:
I think it's fine to start a new chapter - I'm sure we'll have other tidy eval guidelines in the future.

lorenzwalthert replied:
Ok. So if I understand you correctly, we can pull over both new sections I proposed into a new chapter *tidyeval*? Maybe inserting that between the chapter on pipes and the one on code documentation?
Further, should we simplify

> "[...] Note that the sequence of exclamation marks must not be interrupted
> +with spaces, otherwise the meaning is changed to the logical operator `!` [...]"

To

> Don't confuse `!!` and `!!!` with `!`.

?

hadley replied:
Yes, perfect.

lorenzwalthert commented:
I updated the PR according to your suggestions and 

* changed the wording slightly ("one should be explicit about operator precedence" instead of "one
  needs to be explicit about operator precedence"). 
* adapted the example for operator precedence so `# Good` and `# Bad` evaluate to the same.

hadley commented on 05-tidyeval.Rmd:
```diff
@@ -0,0 +1,31 @@
+quo(1 + !! 1 + 2 + 3)
+
+# Good
+quo(1 + (1 + 2 + 3))
```
I think you're missing !! here

And I think for now, let's not promote the use of UQ.

lorenzwalthert commented:
Right... 🙈. I added `!!` to the example and removed the example with and reference to `UQ()`.
